sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java

```diff
@@ -0,0 +1,364 @@
+    // are being accessed. Currently Beam only supports wildcard descriptors.
+    // Once BEAM-4457 is fixed, fix this.
+    @FieldAccess("filterFields")
+    final FieldAccessDescriptor fieldAccessDescriptor = FieldAccessDescriptor.withAllFields();
+
+    public abstract String jdbcUrl();
+
+    public abstract String table();
+
+    public abstract TableSchema schema();
+
+    public abstract Properties properties();
+
+    public static <T> WriteFn<T> create(
+        String jdbcUrl, String table, TableSchema schema, Properties properties) {
+      return new AutoValue_ClickHouseIO_WriteFn<>(jdbcUrl, table, schema, properties);
+    }
+
+    static String quoteIdentifier(String identifier) {
+      String backslash = "\\\\";
+      String quote = "\"";
+
+      return quote + identifier.replaceAll(quote, backslash + quote) + quote;
+    }
+
+    @VisibleForTesting
+    static String insertSql(TableSchema schema, String table) {
+      String columnsStr =
+          schema
+              .columns()
+              .stream()
+              .filter(x -> !x.materializedOrAlias())
+              .map(x -> quoteIdentifier(x.name()))
+              .collect(Collectors.joining(", "));
+      return "INSERT INTO " + quoteIdentifier(table) + " (" + columnsStr + ")";
+    }
+
+    @Setup
+    public void setup() throws SQLException {
+      connection = new ClickHouseDataSource(jdbcUrl(), properties()).getConnection();
+      executor =
+          Executors.newSingleThreadExecutor(
+              new ThreadFactoryBuilder().setNameFormat("clickhouse-jdbc-%d").build());
+      queue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+      bundleFinished = new AtomicBoolean(false);
+    }
+
+    @StartBundle
+    public void startBundle() throws SQLException {
+      statement = connection.createStatement();
+      bundleFinished.set(false);
+
+      // When bundle starts, we open statement and stream data in bundle.
+      // When we finish bundle, we close statement.
+
+      // For streaming, we create a background thread for http client,
+      // and we feed it through BlockingQueue.
+
+      insert =
+          executor.submit(
+              () -> {
+                try {
+                  statement.sendRowBinaryStream(
+                      insertSql(schema(), table()),
+                      stream -> {
+                        while (true) {
+                          Row row;
+                          try {
+                            row = queue.poll(1, TimeUnit.SECONDS);
+                          } catch (InterruptedException e) {
+                            break; // time to go
+                          }
+
+                          if (row == null) {
+                            if (bundleFinished.get() && queue.isEmpty()) {
+                              // bundle is finished and queue is drained, time to go
+                              break;
+                            } else {
+                              // timeout, let's try again
```

Comment:
From my point of view this might be cleaner with take() and interruption on finishBundle, or sending a poison object through queue on finishBundle. Anyway, this is not a point why you asked for my comments, so this is just a suggestion.

Overall regarding queue, I don't see other options for streaming insert with current WriteFn interface now (what I see from annotations). ClickHouseRowBinaryStream had been stable for a long time, so I don't think it is going to disappear unexpectedly. 
The obvious alternative is to collect data in memory and then send the whole batch through jdbc's PreparedStatement.executeBatch() (this is cleaner from interface point of view, but needs memory and may be slower, at least now, since it currently uses tsv as serialization).
I'm not aware of Beam's typical setup and workload, so I'm not sure which option is best in this case.